### ISSUE TITLE
Ability to ignore finalizeProperty tree scan for a given property

### ISSUE THREAD

**Issue by travikk** [immerjs/immer#465]
## 🚀 Feature Proposal

Given the following state tree:

```typescript
const state = {
   prop1: 'hello',
   prop2: 42,
   deepNestedTree: {
      'someKey1': { /* ANOTHER_BIG_OBJECT_HERE */ },
      'someKey2': { /* ANOTHER_BIG_OBJECT_HERE */ },
      'someKey3': { /* ANOTHER_BIG_OBJECT_HERE */ },
      'someKey4': { /* ANOTHER_BIG_OBJECT_HERE */ },
   }
}
```

I would like to be able to tell Immer to **not** scan `deepNestedTree` and each of it's properties and then each properties of those for finalizing, since I know that I never mutate them but rather just create the whole tree from scratch myself (which comes from an API call). 
So in essence, `deepNestedTree` is a read-only property and it's content is never mutated.

## Motivation

Idea here is that `deepNestedTree` could contain hundreds of thousands of records and those records could be quite wide themselves. It takes a lot of time to traverse that tree and for my use-case it is pointless, as it's never mutated.

## Example

According to this https://github.com/immerjs/immer/blob/master/src/immer.js#L312 I can just freeze `state.deepNestedTree` but is that the intended way to do it? Could we perhapse get away with a function that adds a unique symbol to the object to let Immer know it doesn't need to finalize it?

Something like:

```typescript
import {ignoreObject} from 'immer';

produce(state, draft => {
   draft.myBigTree = generateBigTree()
   ignoreObject(Draft.myBigTree)
}, undefined)

```

**Comment by mweststrate** [comment]
As a quick fix downgrading to 4 fixes this. As at that time people were
complaining this _wasnt_ automated :-P.

I think freezing first manually is a perfectly fine trick to solve this
problem. I favor using existing building blocks over extending api surface
for super specific edge cases. Feel free to PR the tip to the docs, for
example on the page about freezing.

**Comment by travikk** [comment]
Freezing first exposes internal implementation to the consumer, meaning I specifically need to know that Immer will skip an object if it's frozen (and frozen isn't an Immer concept on it's own). 

Furthermore, if that internal detail were to change, it's not immediately obvious and discoverable.
Having an API for that just makes it clearer and obvious from an intent point of view.

**Comment by mweststrate** [comment]
Documenting it pins it down as well :) Especially if a test case is added
to verify that we don't break that behavior in the future.

**Comment by travikk** [comment]
So if you use the `immerable` symbol, that's adhearing to your rule of using existing building blocks :)

**Comment by mweststrate** [comment]
I don't get it, immerable does exactly the opposite?

**Comment by travikk** [comment]
What I had in mind was:

Currently, to make an instance of a custom class specifically immerable you need to do:
```typescript
import {immerable} from "immer"

class Foo {
...
}

const x: Foo = new Foo()
x[immerable] = true
```

What I'm suggesting is to allow to make any **plan** object specifically **not** immerable with the following:

```typescript
import {immerable} from "immer"
const x = {
   hello: 'world'
}

x[immerable] = false
```

A quick implementation of that would be something like this:

```javascript

export function isDraftable(value) {
	if (!value) return false
        if (value[DRAFTABLE] === false) return false // Added this check
	return (
		isPlainObject(value) ||
		!!value[DRAFTABLE] ||
		!!value.constructor[DRAFTABLE] ||
		isMap(value) ||
		isSet(value)
	)
}
```

**Comment by travikk** [comment]
@mweststrate I'm happy to make a PR for this change, if I know it's something you'd consider adding. Let me know what you think.

**Comment by mweststrate** [comment]
Ok, that sounds reasonable! Not 100% convinced yet, but feel free to submit a PR, especially if you can prove with some tests that it won't do any scanning or freezing inside that deep tree. I think applications with `[immerable] = false` can be quite confusing and should be considered an anti-pattern, but it does sound like a reasonable escape hatch for very very specfic cases :)

**Comment by mweststrate** [comment]
And apologies for the slow follow-up

**Comment by mweststrate** [comment]
A change in the next major will be that frozen objects, not originating from the draft, will be not be deeply processed anymore, so this means that calling `Object.frozen()` on the big tree should prevent the issue

**Comment by mweststrate** [comment]
Released as 6.0.0

**Comment by cefn** [comment]
Hi, did anything come of this proposal? 

I found myself needing the same thing in the context of Typescript classes, where the use of classes and instanceof is a huge convenience and eliminates duplicate code and probable errors compared to any other strategy. 

I've tried many strategies for my case, I've been cycling around; simple interfaces and explicit object definition, factory functions, class definitions - with each creating their own problems and forcing me to try further iterations of the others. 

I've come to the conclusion the only thing preventing me from adopting classes to manage these structures is the unpredictability of Immer when they are encountered in a state tree. I would be more-than-happy to explicitly flag that runtime errors are expected and even desired when attempting to draft properties of these classes, by setting them `[immerable]:false`. They are effectively 'leaves' of the state tree and while leaves can be moved around in objects and arrays, any edit happening within the leaves should throw.

Using Object.freeze() as part of the constructor definition defeats the use of the classes for inheritance without hacky workarounds like... 

```typescript
class Point3d extends Point2d {
  constructor (x, y, z) {
    super(x, y)
    this.z = z
    if (new.target === Point3d) {
      Object.freeze(this)
    }
  }
}
```
https://kubyshkin.name/posts/immutable-classes-in-javascript/

...a step which cannot be forgotten for any concrete class.

Being able to signal explicitly to immer just once in a common superclass that drafting this structure is an error would be a great benefit and seems intuitively simple and well-aligned with the case.
